clients/web/src/server.js

```diff
@@ -1,50 +1,100 @@
-            wait();
-        });
-    }
+function _retryUntilFulfilled(func, delay) {
+    const resolution = $.Deferred();
+    const loop = () => {
+        // Run the function
+        func()
+            .done(function () {
+                // If it's fulfilled, then fulfill the resolution with the same values
+                resolution.resolve.apply(resolution, arguments);
+            })
+            .fail(() => {
+                // If it's rejected, check again
+                window.setTimeout(loop, delay);
+            });
+    };
+    loop();
+    return resolution.promise();
+}
 
-    return Promise.resolve(restRequest({
+/**
+ * Restart the server, wait until it has restarted, then reload the current page.
+ * @returns {$.Promise} A promise which resolves when the server is restarted.
+ */
+function restartServer() {
+    // Query the server first
+    return restRequest({
         type: 'GET',
         path: 'system/version'
-    }).done(resp => {
-        restartServer._lastStartDate = resp.serverStartDate;
-        restartServer._callSystemRestart();
-        events.trigger('g:alert', {
-            icon: 'cw',
-            text: 'Restarting server',
-            type: 'warning',
-            timeout: 60000
+    })
+        // Restart the server
+        .then((resp) => {
+            // Store the last start date in an attribute, so testing code can mutate it
+            restartServer._lastStartDate = resp.serverStartDate;
+            events.trigger('g:alert', {
+                icon: 'cw',
+                text: 'Restarting server',
+                type: 'warning',
+                timeout: 60000
+            });
+            return restartServer._callSystemRestart();
+        })
+        // Check until the server is restarted
+        .then(() => _retryUntilFulfilled(
+            () => restartServer._checkServer(restartServer._lastStartDate),
+            1000
+        ))
+        // Reload the window after everything completes
+        .done(() => {
+            restartServer._reloadWindow();
         });
-        return waitForServer();
-    }));
 }
 
+/**
+ * Check once whether the server is restarted, returning a promise which reflects the status.
+ * @param {string} lastStartDate A timestamp string with the original start date.
+ * @returns {$.Promise} A promise which resolves or rejects, depending on the restarted state.
+ */
+restartServer._checkServer = function (lastStartDate) {
+    return restRequest({
+        type: 'GET',
+        path: 'system/version',
+        error: null
+    })
+    .then((resp) => {
+        const checkResolution = $.Deferred();
+        if (resp.serverStartDate !== lastStartDate) {
+            checkResolution.resolve();
+        } else {
+            checkResolution.reject();
+        }
+        return checkResolution.promise();
+    });
+};
+
+/* Having these as object properties facilitates testing */
+restartServer._callSystemRestart = function () {
+    restRequest({
```

Comment:
Do we need to be returning this value? I notice above that `restartServer` asynchronously returns the return value from this function, which is always undefined as written.